test/CAPI/firrtl.c

```diff
@@ -0,0 +1,217 @@
+  }};
+  Op firCircuit =
+      buildOp(ctx, mlirModuleGetBody(module), "firrtl.circuit",
+              ARRAY_SIZE(circuitAttrs), circuitAttrs, 0, NULL, 0, NULL,
+              ARRAY_SIZE(circuitBuildRegions), circuitBuildRegions, unkLoc);
+
+  FIRRTLPortDir portsDirs[] = {
+      FIRRTL_PORT_DIR_INPUT,
+  };
+  MlirAttribute portsNames[] = {
+      mlirStringAttrGet(ctx, mlirStringRefCreateFromCString("portIn")),
+  };
+  MlirAttribute portTypes[] = {
+      mlirTypeAttrGet(firrtlTypeGetUInt(ctx, 32)),
+  };
+  MlirAttribute portAnnotations[] = {
+      emptyArrayAttr,
+  };
+  MlirAttribute symAttrs[] = {};
+  MlirAttribute portLocations[] = {
+      mlirLocationGetAttribute(unkLoc),
+  };
+
+  MlirNamedAttribute moduleAttrs[] = {
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx, mlirStringRefCreateFromCString("sym_name")),
+          mlirStringAttrGet(ctx, mlirStringRefCreateFromCString("Circuit"))),
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx, mlirStringRefCreateFromCString("annotations")),
+          emptyArrayAttr),
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx,
+                            mlirStringRefCreateFromCString("portDirections")),
+          firrtlAttrGetPortDirs(ctx, ARRAY_SIZE(portsDirs), portsDirs)),
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx, mlirStringRefCreateFromCString("portNames")),
+          mlirArrayAttrGet(ctx, ARRAY_SIZE(portsNames), portsNames)),
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx, mlirStringRefCreateFromCString("portTypes")),
+          mlirArrayAttrGet(ctx, ARRAY_SIZE(portTypes), portTypes)),
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx,
+                            mlirStringRefCreateFromCString("portAnnotations")),
+          mlirArrayAttrGet(ctx, ARRAY_SIZE(portAnnotations), portAnnotations)),
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx, mlirStringRefCreateFromCString("portSyms")),
+          mlirArrayAttrGet(ctx, ARRAY_SIZE(symAttrs), symAttrs)),
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx,
+                            mlirStringRefCreateFromCString("portLocations")),
+          mlirArrayAttrGet(ctx, ARRAY_SIZE(portLocations), portLocations)),
+  };
+  BuildRegionBlock moduleBuildBlock = {
+      .numArgs = 0,
+      .args = NULL,
+      .locs = NULL,
+  };
+  BuildRegion moduleBuildRegions[] = {{
+      .numBlocks = 1,
+      .blocks = &moduleBuildBlock,
+  }};
+  Op firModule =
+      buildOp(ctx, firCircuit.regions[0].blocks[0], "firrtl.module",
+              ARRAY_SIZE(moduleAttrs), moduleAttrs, 0, NULL, 0, NULL,
+              ARRAY_SIZE(moduleBuildRegions), moduleBuildRegions, unkLoc);
+
+  assert(mlirLogicalResultIsSuccess(
+      mlirExportFIRRTL(module, exportCallback, NULL)));
+
+  // CHECK: FIRRTL version 3.0.0
+  // CHECK: circuit Circuit :
+  // CHECK:   module Circuit :
+  // CHECK:     input portIn : UInt<32>
```

Comment:
Nits: `CHECK-NEXT` for these is a tighter check.

Reply:
Done.